### ISSUE TITLE
Wallpaper uses too much ram

### ISSUE THREAD

**Issue by arte921** [swaywm/swaybg#14]
I'm using sway version 1.0 on Fedora 31. I use the version from the default repos. I obtained my high resolution wallpaper from https://i.redd.it/ijsf9nney0a41.jpg using this exact command, just after a fresh install:

`curl https://i.redd.it/ijsf9nney0a41.jpg > /home/arte/pictures/ijsf9nney0a41.jpg`

I enabled the wallpaper in my sway config using this line:

`output * bg /home/arte/pictures/ijsf9nney0a41.jpg fill`

When I now boot and sway automatically starts(because I put "sway" in .bash_profile) my ram usage is over 1.5 GB. top shows this is mainly caused by a 1268 MB large process called swaybg. the full command is, according to cat  `/proc/1025/cmdline` where 1025 is the result of `pidof swaybg`:

swaybg-oHDMI-A-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-oDVI-D-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-oDP-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-o*-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill

The wallpaper shows normally, altough it takes a few seconds after starting sway before it shows up.

I couldn't attach it on github so my sway config from `~/.config/sway/config` is at [https://raw.githubusercontent.com/arte921/sway-configs/master/pc](https://raw.githubusercontent.com/arte921/sway-configs/master/pc)

**Comment by kmARC** [comment]
Also noticed it. Fresh start with some default gnome background pixmap eats up 400M RES

![image](https://user-images.githubusercontent.com/6640417/79670894-69fe4280-81c6-11ea-8f66-31b776ac7e3a.png)

**Comment by pseregiet** [comment]
sway will load your picture and decode it from png or other format into the raw pixel data (4 bytes per pixel). Your image is 9000x9000 which is 81000000 bytes (77mb).

**Comment by Logarithmus** [comment]
> I'm using sway version 1.0 on Fedora 31. I use the version from the default repos. I obtained my high resolution wallpaper from https://i.redd.it/ijsf9nney0a41.jpg using this exact command, just after a fresh install:
> 
> `curl https://i.redd.it/ijsf9nney0a41.jpg > /home/arte/pictures/ijsf9nney0a41.jpg`
> 
> I enabled the wallpaper in my sway config using this line:
> 
> `output * bg /home/arte/pictures/ijsf9nney0a41.jpg fill`
> 
> When I now boot and sway automatically starts(because I put "sway" in .bash_profile) my ram usage is over 1.5 GB. top shows this is mainly caused by a 1268 MB large process called swaybg. the full command is, according to cat `/proc/1025/cmdline` where 1025 is the result of `pidof swaybg`:
> 
> swaybg-oHDMI-A-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-oDVI-D-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-oDP-1-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill-o*-i/home/arte/pictures/ijsf9nney0a41.jpg-mfill
> 
> The wallpaper shows normally, altough it takes a few seconds after starting sway before it shows up.
> 
> I couldn't attach it on github so my sway config from `~/.config/sway/config` is at https://raw.githubusercontent.com/arte921/sway-configs/master/pc

Just downscale it to your screen resolution (e. g. 1920x1080):
`vipsthumbnail --smartcrop attention -s 1920x1080 /home/arte/pictures/ijsf9nney0a41.jpg`
The downscaled version will be written to `/home/arte/pictures/tn_ijsf9nney0a41.jpg`

**Comment by mstoeckl** [comment]
> Your image is 9000x9000 which is 81000000 bytes (77mb).

I had some time to go over the memory consumption of swaybg in more detail. On a command like:
```
swaybg -o HDMI-A-1 -i /home/arte/pictures/ijsf9nney0a41.jpg -mfill -o DVI-D-1 -i /home/arte/pictures/ijsf9nney0a41.jpg -m fill -o DP-1 -i /home/arte/pictures/ijsf9nney0a41.jpg -mfill -o '*' -i /home/arte/pictures/ijsf9nney0a41.jpg -m fill
```
for each of the four output configurations created (HDMI-A-1, DVI-D-1, DP-1, '*'):
* swaybg loads the jpg file with gdk_pixbuf_new_from_file. Because jpg files have no alpha channel, they are stored using 24 bits per pixel/3 bytes each; this uses 243e6 bytes (scientific notation) in total.
* The gdk_pixbuf is converted to an RGB24 cairo surface, using 4 bytes per pixel, for 324e6 bytes in total; then the gdk_pixbuf is deleted. (While the gdk_pixbuf doesn't affect long term memory usage, it does increase peak memory usage, which can make swapping on startup more likely with huge images.)
* Each swaybg output has two slots for a shm buffer to use for its background image; because computer monitors are not resizeable, only one of these buffers is used. The shm buffers use 4 bytes per pixel and have format WL_SHM_FORMAT_ARGB8888. A 1920x1080 monitor, for example, would use 8.3e6 bytes.
* Finally, sway itself copies each shm buffer into an OpenGL texture; the texture is, I think, also stored as DRM_FORMAT_ARGB8888, and uses 4 bytes per pixel. This would only count towards sway's memory usage.

Based on this, we'd expect `324e6*4 + 8.29e6*4 = 1329.2e6` bytes of memory usage. The original poster reports 1268 MB = `1268*1.048e6 = 1329.6e6` bytes of memory usage.

Ironically, the only memory here that is needed after setting up the final images is that from the OpenGL textures, which count toward the memory use of the sway process.

I have an experimental branch of swaybg which loads all images on demand, and doesn't cache the large cairo surfaces; see https://github.com/mstoeckl/swaybg/commits/low-memory . The tradeoff for the memory savings is significantly slower background redrawing when running sway nested.

**Comment by emersion** [comment]
Because swaybg doesn't nées to draw more than once, we coudl adopt a strategy to release buffers on `wl_buffer.release`. This would result in a zero memory usage by the swaybg process once the OpenGL texture is uploaded.

**Comment by mstoeckl** [comment]
> Because swaybg doesn't need to draw more than once, we coudl adopt a strategy to release buffers on wl_buffer.release. This would result in a zero memory usage by the swaybg process once the OpenGL texture is uploaded.

Since the buffers would then only be used once, they can be destroyed on the swaybg side immediately after sending wl_surface.commit. I've updated the low-memory branch to do this.

**Comment by emersion** [comment]
Oh, right, that's even better. Ref: https://gitlab.freedesktop.org/wayland/wayland/-/issues/185

**Comment by emersion** [comment]
@mstoeckl, any plans to submit your low-memory branch as a PR?

**Comment by mstoeckl** [comment]
The low-memory branch does some redundant image loading, and there's a better way to load images with gdk_pixbuf ; once I have these issues fixed (for which I shall probably have time before the end of June) I will make a PR.

**Comment by mstoeckl** [comment]
What remains to be done after #27 is:
* Using GdkPixbufLoader to load and scale images in one operation. (This significantly reduce memory usage and startup time for huge JPG backgrounds.) I will open a PR for this soon.
* More fine grained buffer management (i.e., tracking buffer release operations, and reusing existing buffers instead of rendering a new one when an output is dirty but unchanged, or if two outputs have exactly matching images and sizes.)

**Comment by emersion** [comment]
#27 already improves the situation a lot, so I'll close this. If we want to go further, we'll probably want to stop using GdkPixbuf, as noted in https://github.com/swaywm/swaybg/pull/28#issuecomment-876835782.

If you're interested in working on this, please open a new issue so that we can discuss about potential alternatives like GEGL.

**Comment by cab404** [comment]
Not closed @emersion 
![image](https://user-images.githubusercontent.com/6453661/156758765-70a6f905-945f-45e3-b76e-94f0d11592f7.png)

**Comment by emersion** [comment]
@cab404 https://github.com/swaywm/swaybg/pull/40 may help
